### ISSUE TITLE
Rethinking MVKDevice::markLost()?

### ISSUE THREAD

**Issue by billhollings** [KhronosGroup/MoltenVK#1366]
Currently `MVKDevice::markLost()` signals a fence (and semaphores) that is also later signalled in the same `MTLCommandBuffer` completion handler within `MVKQueueCommandBufferSubmission::finish()`.

A problem arises when the code waiting on the fence signal is triggered by `markLost()`, and then destroys the fence (either during app flow, or when `mvkWaitForFences()`finishes) _**before**_ the `MVKQueueCommandBufferSubmission::finish()` is called in the `MTLCommandBuffer` completion handler, which then results in a memory access error.

At the very least, `MVKQueueCommandBufferSubmission` needs to `retain()/release()` the fence and wait semaphores to avoid this race condition.

But in reviewing this, I also started considering whether `MVKDevice::markLost()` should really be always signalling everything that is currently waiting on the `MVKDevice`. Many `MTLCommandBuffer` errors are local to that submission (eg. a bad resource), and the fence and semaphores for that submission will be signalled anyway in `MVKQueueCommandBufferSubmission::finish()`.

1. Is there really a need to signal all the other sync objects waiting on other threads? Will not those also be signalled by each respective `MVKQueueCommandBufferSubmission::finish()` being called?

2. And if there is a need (maybe for events?), should it only be done for `MTLCommandBuffer` error enums that truly indicate a device loss?

3. And in that case, should the collection of waiting syncs by managed within `MVKPhysicalDevice`, instead of `MVKDevice`?

@cdavis5e

**Comment by billhollings** [comment]
I'll take a stab at a PR for this. I'll focus on:

- Adding `retain()/release()` for the `MVKQueueCommandBufferSubmission` fence and signal semaphores.
- Moving `markLost()` and the collection of waiting syncs to `MVKPhysicalDevice`, and only invoking it if the error indicates the entire `MTLDevice` is compromised, or if `MVKConfiguration::resumeLostDevice` is disabled.
- Enabling `MVKConfiguration::resumeLostDevice` by default.

**Comment by billhollings** [comment]
PR #1367 fixes the race condition.

I decided not to do anything further about moving `markLost()` to `MVKPhysicalDevice`. The Vulkan spec is clear that the app should destroy and re-create a `VkDevice` under these conditions, so the current behavior is correct (even if the spec is heavy-handed). 

I did change the behavior under `MVKConfiguration::resumeLostDevice`, so that the waiting semaphores are not signalled unless it's a `VkPhysicalDevice` fatal error.

Closing.
